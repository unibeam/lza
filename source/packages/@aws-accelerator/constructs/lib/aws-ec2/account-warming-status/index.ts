/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
/**
 * aws-ec2-pre-warm-account-status - lambda handler
 *
 * @param event
 * @returns
 */

import { SSMClient, PutParameterCommand } from '@aws-sdk/client-ssm';
import {
  EC2Client,
  DescribeInstancesCommand,
  DescribeVpcsCommand,
  DescribeSubnetsCommand,
  DeleteVpcCommand,
  DeleteSubnetCommand,
  TerminateInstancesCommand,
} from '@aws-sdk/client-ec2';
import { AdaptiveRetryStrategy } from '@aws-sdk/util-retry';
import { delay } from '@aws-accelerator/utils/lib/throttle';

const solutionId = process.env['SOLUTION_ID'] ?? '';
const retryStrategy = new AdaptiveRetryStrategy(() => Promise.resolve(5));
const ec2Client = new EC2Client({ customUserAgent: solutionId, retryStrategy });
const ssmClient = new SSMClient({ customUserAgent: solutionId, retryStrategy });

type InstanceDetails = {
  instanceId: string | undefined;
  minutesSinceLaunch: number | undefined;
};

//eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function handler(event: any): Promise<
  | {
      IsComplete: boolean;
    }
  | undefined
> {
  console.log(event);
  const ssmPrefix: string = event.ResourceProperties['ssmPrefix'];
  const instanceDetails = await getInstanceDetails();

  // if no instance was found, it has been deleted return
  if (instanceDetails.instanceId === undefined) {
    return { IsComplete: true };
  }

  if (instanceDetails.minutesSinceLaunch! > 15) {
    console.log('Account warming time reached');
    await terminateInstance(instanceDetails.instanceId);
    await deleteVpc();
    await setSSMParameter(ssmPrefix, 'true');
    return {
      IsComplete: true,
    };
  }
  console.log('Waiting for warmup to complete');
  return { IsComplete: false };
}

async function getInstanceDetails(): Promise<InstanceDetails> {
  console.log('Getting Instance Id');
  const instanceDetails: InstanceDetails = { instanceId: undefined, minutesSinceLaunch: undefined };
  const ec2Instances = await ec2Client.send(
    new DescribeInstancesCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (ec2Instances.Reservations && ec2Instances.Reservations.length < 1) {
    return instanceDetails;
  }

  if (ec2Instances.Reservations![0].Instances?.length ?? 0 > 0) {
    instanceDetails.instanceId = ec2Instances.Reservations![0].Instances![0].InstanceId;
    const launchTime = ec2Instances.Reservations![0].Instances![0].LaunchTime;

    const currentTime = new Date();
    const msElapsed = currentTime.getTime() - launchTime!.getTime();
    instanceDetails.minutesSinceLaunch = msElapsed / 1000 / 60;
  }

  return instanceDetails;
}

async function deleteVpc() {
  console.log('Deleting VPC');
  const vpcId = await getVpcId();

  if (vpcId) {
    console.log('Deleting subnets');
    const subnets = await ec2Client.send(
      new DescribeSubnetsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }),
    );
    for (const subnet of subnets.Subnets) {
      await ec2Client.send(new DeleteSubnetCommand({ SubnetId: subnet.SubnetId }));
    }
    console.log(`Deleting VPC with id: ${vpcId}`);
    await ec2Client.send(new DeleteVpcCommand({ VpcId: vpcId }));
  }
}

async function getVpcId(): Promise<string | undefined> {
  console.log('Getting VPC Id');
  const vpcs = await ec2Client.send(
    new DescribeVpcsCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (vpcs.Vpcs?.length ?? 0 > 0) {
    return vpcs.Vpcs![0].VpcId;
  }
  return undefined;
}

async function setSSMParameter(ssmPrefix: string, parameterValue: string) {
  console.log('Updating SSM Parameter');
  try {
    await ssmClient.send(
      new PutParameterCommand({
        Name: `${ssmPrefix}/account/pre-warmed`,
        Value: parameterValue,
        Overwrite: true,
      }),
    );
  } catch (e) {
    console.log(e);
  }
}

async function terminateInstance(instanceId: string): Promise<void> {
  console.log(`Terminating EC2 instance: ${instanceId}`);
  await ec2Client.send(new TerminateInstancesCommand({ InstanceIds: [instanceId] }));
  await waitForTermination(instanceId);
}

async function waitForTermination(instanceId: string): Promise<void> {
  console.log(`Waiting for termination of instanceId: ${instanceId}`);
  let ec2Terminated = false;
  while (!ec2Terminated) {
    const statusResponse = await ec2Client.send(new DescribeInstancesCommand({ InstanceIds: [instanceId] }));
    if (
      statusResponse.Reservations[0].Instances[0].State?.Name !== 'terminated' ||
      statusResponse.Reservations[0].Instances[0].State?.Code !== 48
    ) {
      delay(15000);
      continue;
    }
    ec2Terminated = true;
  }
}
