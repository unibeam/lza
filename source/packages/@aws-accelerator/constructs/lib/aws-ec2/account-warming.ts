/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import path = require('path');

export interface WarmAccountProps {
  readonly cloudwatchKmsKey?: cdk.aws_kms.IKey;
  readonly logRetentionInDays: number;
  readonly ssmPrefix: string;
}

export class WarmAccount extends Construct {
  readonly onEvent: cdk.aws_lambda.IFunction;
  readonly isComplete: cdk.aws_lambda.IFunction;
  readonly provider: cdk.custom_resources.Provider;
  readonly id: string;

  constructor(scope: Construct, id: string, props: WarmAccountProps) {
    super(scope, id);

    const WARM_ACCOUNT = 'Custom::WarmAccount';

    const ec2Policy = new cdk.aws_iam.PolicyStatement({
      sid: 'ec2',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:DescribeVpcs', 'ec2:DescribeInstances', 'ec2:DescribeSubnets'],
      resources: ['*'],
    });

    const ec2DeletePolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'ec2Delete',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:DeleteVpc', 'ec2:DeleteSubnet', 'ec2:TerminateInstances'],
      resources: ['*'],
      conditions: {
        StringEquals: {
          'aws:ResourceTag/Name': 'accelerator-warm',
        },
      },
    });

    const ec2RunPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'ec2Run',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:RunInstances'],
      resources: [
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:subnet/*`,
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:instance/*`,
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:volume/*`,
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${
          cdk.Stack.of(this).account
        }:security-group/*`,
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}::image/*`,
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}::snapshot/*`,
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${
          cdk.Stack.of(this).account
        }:network-interface/*`,
      ],
    });

    const ec2CreateTags = new cdk.aws_iam.PolicyStatement({
      sid: 'ec2CreateTags',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:CreateTags'],
      resources: [
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:instance/*`,
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:volume/*`,
      ],
      conditions: {
        StringEquals: {
          'ec2:CreateAction': 'RunInstances',
        },
      },
    });

    const ec2CreateVpcTags = new cdk.aws_iam.PolicyStatement({
      sid: 'ec2CreateVpcTags',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:CreateTags'],
      resources: [
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:vpc/*`,
      ],
      conditions: {
        StringEquals: {
          'ec2:CreateAction': 'CreateVpc',
        },
      },
    });

    const ec2CreateSubnetTags = new cdk.aws_iam.PolicyStatement({
      sid: 'ec2CreateSubnetTags',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:CreateTags'],
      resources: [
        `arn:${cdk.Stack.of(this).partition}:ec2:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:subnet/*`,
      ],
      conditions: {
        StringEquals: {
          'ec2:CreateAction': 'CreateSubnet',
        },
      },
    });

    const ec2CreationPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'vpccreation',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:CreateVpc', 'ec2:CreateSubnet'],
      resources: ['*'],
    });

    const ssmPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'ssm',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ssm:GetParameter', 'ssm:PutParameter', 'ssm:DeleteParameter'],
      resources: [
        `arn:${cdk.Stack.of(this).partition}:ssm:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:parameter${
          props.ssmPrefix
        }/account/pre-warmed`,
      ],
    });

    const ssmAmiPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'ssmami',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ssm:GetParameter'],
      resources: [
        `arn:${cdk.Stack.of(this).partition}:ssm:${
          cdk.Stack.of(this).region
        }::parameter/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2`,
      ],
    });

    this.onEvent = new cdk.aws_lambda.Function(this, 'WarmAccountFunction', {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'account-warming/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      timeout: cdk.Duration.minutes(5),
      description: 'Account warming onEvent handler',
      initialPolicy: [
        ec2Policy,
        ec2DeletePolicy,
        ec2CreationPolicy,
        ec2RunPolicy,
        ec2CreateTags,
        ec2CreateVpcTags,
        ec2CreateSubnetTags,
        ssmPolicy,
        ssmAmiPolicy,
      ],
    });

    const onEventLogGroup = new cdk.aws_logs.LogGroup(this, `${this.onEvent.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${this.onEvent.functionName}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.cloudwatchKmsKey,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
    });

    this.isComplete = new cdk.aws_lambda.Function(this, 'WarmAccountStatusFunction', {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'account-warming-status/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      timeout: cdk.Duration.minutes(5),
      description: 'Account warming isComplete handler',
      initialPolicy: [ec2Policy, ec2DeletePolicy, ssmPolicy],
    });

    const isCompleteLogGroup = new cdk.aws_logs.LogGroup(this, `${this.isComplete.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${this.isComplete.functionName}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.cloudwatchKmsKey,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
    });
    const waiterStateMachineLogGroup = new cdk.aws_logs.LogGroup(this, `${this.onEvent.node.id}WaiterLogGroup`, {
      logGroupName: `/aws/vendedlogs/states/waiter-state-machine/${this.onEvent.node.id}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.cloudwatchKmsKey,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    this.provider = new cdk.custom_resources.Provider(this, 'WarmAccountProvider', {
      onEventHandler: this.onEvent,
      isCompleteHandler: this.isComplete,
      queryInterval: cdk.Duration.minutes(2),
      totalTimeout: cdk.Duration.hours(1),
      waiterStateMachineLogOptions: {
        destination: waiterStateMachineLogGroup,
        includeExecutionData: true,
        level: cdk.aws_stepfunctions.LogLevel.ERROR, // error is the default level that CDK auto-creates
      },
    });

    const resource = new cdk.CustomResource(this, 'WarmAccountResource', {
      resourceType: WARM_ACCOUNT,
      serviceToken: this.provider.serviceToken,
      properties: {
        ssmPrefix: props.ssmPrefix,
      },
    });

    NagSuppressions.addResourceSuppressions(
      this.isComplete,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'Custom resource lambda require access to other services',
        },
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Policy is limited to certain tag properties',
        },
      ],
      true,
    );

    NagSuppressions.addResourceSuppressions(
      this.onEvent,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'Custom resource lambda require access to other services',
        },
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Policy is limited to certain tag properties',
        },
      ],
      true,
    );

    NagSuppressions.addResourceSuppressions(
      this.provider,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'CDK Generated Role',
        },
        {
          id: 'AwsSolutions-IAM5',
          reason: 'CDK Generated Policy',
        },
        {
          id: 'AwsSolutions-SF1',
          reason: 'CDK Generated StateMachine',
        },
        {
          id: 'AwsSolutions-SF2',
          reason: 'CDK Generated StateMachine',
        },
      ],
      true,
    );
    // Ensure that the LogGroup is created by Cloudformation prior to Lambda execution
    resource.node.addDependency(isCompleteLogGroup);
    resource.node.addDependency(onEventLogGroup);
    resource.node.addDependency(waiterStateMachineLogGroup);
    this.id = resource.ref;
  }
}
