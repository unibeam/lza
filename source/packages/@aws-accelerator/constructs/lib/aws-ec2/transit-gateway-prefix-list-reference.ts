/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

export interface TransitGatewayPrefixListReferenceProps {
  /**
   * The prefix list ID to reference
   */
  readonly prefixListId: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly logGroupKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
  /**
   * The ID of the transit gateway route table.
   */
  readonly transitGatewayRouteTableId: string;
  /**
   * Determines if route is blackholed.
   */
  readonly blackhole?: boolean;
  /**
   * Owning account ID for cross-account TGW associations
   */
  readonly owningAccountId?: string;
  /**
   * Owning region for cross-account TGW associations
   */
  readonly owningRegion?: string;
  /**
   * Role name for cross-account TGW associations
   */
  readonly roleName?: string;
  /**
   * The identifier of the Transit Gateway Attachment
   *
   */
  readonly transitGatewayAttachmentId?: string;
}

export class TransitGatewayPrefixListReference extends cdk.Resource {
  public readonly id: string;

  constructor(scope: Construct, id: string, props: TransitGatewayPrefixListReferenceProps) {
    super(scope, id);

    const policyStatements = [
      {
        Sid: 'AllowModifyTgwReferences',
        Effect: 'Allow',
        Action: [
          'ec2:CreateTransitGatewayPrefixListReference',
          'ec2:ModifyTransitGatewayPrefixListReference',
          'ec2:DeleteTransitGatewayPrefixListReference',
        ],
        Resource: '*',
      },
    ];

    if (props.roleName) {
      policyStatements.push({
        Sid: 'AssumeRole',
        Effect: 'Allow',
        Action: ['sts:AssumeRole'],
        Resource: `arn:${this.stack.partition}:iam::*:role/${props.roleName}`,
      });
    }

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, 'Custom::TransitGatewayPrefixListReference', {
      codeDirectory: path.join(__dirname, 'transit-gateway-prefix-list-reference/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements,
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: 'Custom::TransitGatewayPrefixListReference',
      serviceToken: provider.serviceToken,
      properties: {
        prefixListReference: {
          PrefixListId: props.prefixListId,
          TransitGatewayRouteTableId: props.transitGatewayRouteTableId,
          Blackhole: props.blackhole,
          TransitGatewayAttachmentId: props.transitGatewayAttachmentId,
        },
        owningAccountId: props.owningAccountId,
        owningRegion: props.owningRegion,
        roleName: props.roleName,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.logGroupKmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
