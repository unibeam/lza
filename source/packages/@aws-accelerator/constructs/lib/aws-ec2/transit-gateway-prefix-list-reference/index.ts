/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/**
 * aws-ec2-transit-gateway-prefix-list-reference - lambda handler
 *
 * @param event
 * @returns
 */

import {
  CreateTransitGatewayPrefixListReferenceCommand,
  DeleteTransitGatewayPrefixListReferenceCommand,
  EC2Client,
  ModifyTransitGatewayPrefixListReferenceCommand,
} from '@aws-sdk/client-ec2';

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { AssumeRoleCommand, STSClient } from '@aws-sdk/client-sts';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
interface ReferenceOptions {
  /**
   * API props
   */
  readonly props: {
    readonly PrefixListId: string;
    readonly TransitGatewayRouteTableId: string;
    readonly Blackhole?: boolean;
    readonly TransitGatewayAttachmentId?: string;
  };
  /**
   * Invoking account ID for the custom resource
   */
  readonly invokingAccountId: string;
  /**
   * Invoking region for the custom resource
   */
  readonly invokingRegion: string;
  /**
   * Custom resource partition
   */
  readonly partition: string;
  /**
   * Owning account ID for cross-account customer gateways
   */
  readonly owningAccountId?: string;
  /**
   * Owning region for cross-account customer gateways
   */
  readonly owningRegion?: string;
  /**
   * Role name for cross-account customer gateways
   */
  readonly roleName?: string;
}

export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
    }
  | undefined
> {
  const options = setOptions(event.ResourceProperties, event.ServiceToken);
  const ec2Client = await setEc2Client(options, process.env['SOLUTION_ID']);

  switch (event.RequestType) {
    case 'Create':
      await throttlingBackOff(() => ec2Client.send(new CreateTransitGatewayPrefixListReferenceCommand(options.props)));

      return {
        Status: 'SUCCESS',
      };

    case 'Update':
      await throttlingBackOff(() => ec2Client.send(new ModifyTransitGatewayPrefixListReferenceCommand(options.props)));

      return {
        Status: 'SUCCESS',
      };

    case 'Delete':
      await throttlingBackOff(() =>
        ec2Client.send(
          new DeleteTransitGatewayPrefixListReferenceCommand({
            PrefixListId: options.props.PrefixListId,
            TransitGatewayRouteTableId: options.props.TransitGatewayRouteTableId,
          }),
        ),
      );

      return {
        Status: 'SUCCESS',
      };
  }
}

/**
 * Set TGW prefix list reference options based on event
 * @param resourceProperties { [key: string]: any }
 * @param serviceToken string
 * @returns ReferenceOptions
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function setOptions(resourceProperties: { [key: string]: any }, serviceToken: string): ReferenceOptions {
  return {
    props: {
      PrefixListId: resourceProperties['prefixListReference']['PrefixListId'],
      TransitGatewayRouteTableId: resourceProperties['prefixListReference']['TransitGatewayRouteTableId'],
      Blackhole: resourceProperties['prefixListReference']['Blackhole'] === 'true',
      TransitGatewayAttachmentId: resourceProperties['prefixListReference']['TransitGatewayAttachmentId'],
    },
    invokingAccountId: serviceToken.split(':')[4],
    invokingRegion: serviceToken.split(':')[3],
    partition: serviceToken.split(':')[1],
    owningAccountId: (resourceProperties['owningAccountId'] as string) ?? undefined,
    owningRegion: (resourceProperties['owningRegion'] as string) ?? undefined,
    roleName: (resourceProperties['roleName'] as string) ?? undefined,
  };
}

/**
 * Returns a local or cross-account/cross-region EC2 client based on input parameters
 * @param options ReferenceOptions
 * @param solutionId string | undefined
 * @returns Promise<EC2Client>
 */
async function setEc2Client(options: ReferenceOptions, solutionId?: string): Promise<EC2Client> {
  const roleArn = `arn:${options.partition}:iam::${options.owningAccountId}:role/${options.roleName}`;
  const stsClient = new STSClient({ region: options.invokingRegion, customUserAgent: solutionId });

  if (options.owningAccountId && options.owningRegion) {
    if (!options.roleName) {
      throw new Error(`Cross-account TGW prefix list reference required but roleName parameter is undefined`);
    }
    //
    // Assume role via STS
    const credentials = await getStsCredentials(stsClient, roleArn);
    //
    // Return EC2 client
    return new EC2Client({
      region: options.owningRegion,
      customUserAgent: solutionId,
      credentials,
    });
  } else if (options.owningAccountId && !options.owningRegion) {
    if (!options.roleName) {
      throw new Error(`Cross-account TGW prefix list reference required but roleName parameter is undefined`);
    }
    //
    // Assume role via STS
    const credentials = await getStsCredentials(stsClient, roleArn);
    //
    // Return EC2 client
    return new EC2Client({
      region: options.invokingRegion,
      customUserAgent: solutionId,
      credentials,
    });
  } else {
    return new EC2Client({
      region: options.owningRegion ?? options.invokingRegion,
      customUserAgent: solutionId,
    });
  }
}

/**
 * Returns STS credentials for a given role ARN
 * @param stsClient STSClient
 * @param roleArn string
 * @returns `Promise<{ accessKeyId: string; secretAccessKey: string; sessionToken: string }>`
 */
async function getStsCredentials(
  stsClient: STSClient,
  roleArn: string,
): Promise<{ accessKeyId: string; secretAccessKey: string; sessionToken: string }> {
  console.log(`Assuming role ${roleArn}...`);
  try {
    const response = await throttlingBackOff(() =>
      stsClient.send(new AssumeRoleCommand({ RoleArn: roleArn, RoleSessionName: 'AcceleratorAssumeRole' })),
    );
    //
    // Validate response
    if (!response.Credentials?.AccessKeyId) {
      throw new Error(`Access key ID not returned from AssumeRole command`);
    }
    if (!response.Credentials.SecretAccessKey) {
      throw new Error(`Secret access key not returned from AssumeRole command`);
    }
    if (!response.Credentials.SessionToken) {
      throw new Error(`Session token not returned from AssumeRole command`);
    }

    return {
      accessKeyId: response.Credentials.AccessKeyId,
      secretAccessKey: response.Credentials.SecretAccessKey,
      sessionToken: response.Credentials.SessionToken,
    };
  } catch (e) {
    throw new Error(`Could not assume role: ${e}`);
  }
}
