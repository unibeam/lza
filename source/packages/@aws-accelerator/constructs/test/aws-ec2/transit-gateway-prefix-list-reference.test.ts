/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { TransitGatewayPrefixListReference } from '../../lib/aws-ec2/transit-gateway-prefix-list-reference';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(TransitGatewayPrefixListReference): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new TransitGatewayPrefixListReference(stack, 'TestTransitGatewayPrefixListReference', {
  prefixListId: 'pl-test',
  transitGatewayAttachmentId: 'tgw-attach-test',
  transitGatewayRouteTableId: 'Test',
  logGroupKmsKey: new cdk.aws_kms.Key(stack, 'TestKms', {}),
  logRetentionInDays: 3653,
});

/**
 * Transit gateway prefix list reference construct test
 */
describe('TransitGatewayPrefixListReference', () => {
  snapShotTest(testNamePrefix, stack);
});
