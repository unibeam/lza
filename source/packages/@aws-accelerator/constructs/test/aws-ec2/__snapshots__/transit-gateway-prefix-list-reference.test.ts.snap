// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitGatewayPrefixListReference Construct(TransitGatewayPrefixListReference):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomTransitGatewayPrefixListReferenceCustomResourceProviderHandler9BAD63E3": {
      "DependsOn": [
        "CustomTransitGatewayPrefixListReferenceCustomResourceProviderRoleC5D4C080",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomTransitGatewayPrefixListReferenceCustomResourceProviderRoleC5D4C080",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomTransitGatewayPrefixListReferenceCustomResourceProviderLogGroupBC9F3669": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "TestKms67078DF1",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "CustomTransitGatewayPrefixListReferenceCustomResourceProviderHandler9BAD63E3",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomTransitGatewayPrefixListReferenceCustomResourceProviderRoleC5D4C080": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ec2:CreateTransitGatewayPrefixListReference",
                    "ec2:ModifyTransitGatewayPrefixListReference",
                    "ec2:DeleteTransitGatewayPrefixListReference",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "AllowModifyTgwReferences",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestKms67078DF1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TestTransitGatewayPrefixListReference11CAF048": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomTransitGatewayPrefixListReferenceCustomResourceProviderLogGroupBC9F3669",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomTransitGatewayPrefixListReferenceCustomResourceProviderHandler9BAD63E3",
            "Arn",
          ],
        },
        "prefixListReference": {
          "PrefixListId": "pl-test",
          "TransitGatewayAttachmentId": "tgw-attach-test",
          "TransitGatewayRouteTableId": "Test",
        },
      },
      "Type": "Custom::TransitGatewayPrefixListReference",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
